### PR TITLE
Add mutexes in BlockStorage

### PR DESCRIPTION
**Pull request by KaustubhShamshery**
## Description

Add mutex to avoid race conditions in libPersistence

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libPersistence/BlockStorage.cpp [hunk 1]
```diff
@@ -170,6 +195,8 @@ bool BlockStorage::GetRangeMicroBlocks(const uint64_t lowEpochNum,
                                        list<MicroBlockSharedPtr>& blocks) {
   LOG_MARKER();
 
+  shared_lock<shared_timed_mutex> g(m_mutexMicroBlock);
+
   leveldb::Iterator* it =
       m_microBlockDB->GetDB()->NewIterator(leveldb::ReadOptions());
   for (it->SeekToFirst(); it->Valid(); it->Next()) {
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexMicroBlock` to protect `leveldb::Iterator* it = m_microBlockDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 2]
```diff
@@ -217,15 +244,15 @@ bool BlockStorage::PutTempState(const unordered_map<Address, Account>& states) {
...
 
 bool BlockStorage::GetTempStateInBatch(leveldb::Iterator*& iter,
                                        vector<StateSharedPtr>& states) {
   // LOG_MARKER();
 
-  lock_guard<mutex> g(m_mutexTempState);
+  shared_lock<shared_timed_mutex> g(m_mutexTempState);
 
   if (iter == nullptr) {
     iter = m_tempStateDB->GetDB()->NewIterator(leveldb::ReadOptions());
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexTempState` to protect `iter = m_tempStateDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 3]
```diff
@@ -425,6 +500,8 @@ bool BlockStorage::DeleteTxBody(const dev::h256& key) {
 bool BlockStorage::GetAllDSBlocks(std::list<DSBlockSharedPtr>& blocks) {
   LOG_MARKER();
 
+  shared_lock<shared_timed_mutex> g(m_mutexDsBlockchain);
+
   leveldb::Iterator* it =
       m_dsBlockchainDB->GetDB()->NewIterator(leveldb::ReadOptions());
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexDsBlockchain` to protect `leveldb::Iterator* it = m_dsBlockchainDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 4]
```diff
@@ -455,6 +532,8 @@ bool BlockStorage::GetAllDSBlocks(std::list<DSBlockSharedPtr>& blocks) {
 bool BlockStorage::GetAllTxBlocks(std::list<TxBlockSharedPtr>& blocks) {
   LOG_MARKER();
 
+  shared_lock<shared_timed_mutex> g(m_mutexTxBlockchain);
+
   leveldb::Iterator* it =
       m_txBlockchainDB->GetDB()->NewIterator(leveldb::ReadOptions());
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexTxBlockchain` to protect `leveldb::Iterator* it = m_txBlockchainDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 5]
```diff
@@ -491,6 +570,8 @@ bool BlockStorage::GetAllTxBodiesTmp(std::list<TxnHash>& txnHashes) {
 
   LOG_MARKER();
 
+  shared_lock<shared_timed_mutex> g(m_mutexTxBodyTmp);
+
   leveldb::Iterator* it =
       m_txBodyTmpDB->GetDB()->NewIterator(leveldb::ReadOptions());
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexTxBodyTmp` to protect `leveldb::Iterator* it = m_txBodyTmpDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 6]
```diff
@@ -510,6 +591,9 @@ bool BlockStorage::GetAllTxBodiesTmp(std::list<TxnHash>& txnHashes) {
 
 bool BlockStorage::GetAllBlockLink(std::list<BlockLink>& blocklinks) {
   LOG_MARKER();
+
+  shared_lock<shared_timed_mutex> g(m_mutexBlockLink);
+
   leveldb::Iterator* it =
       m_blockLinkDB->GetDB()->NewIterator(leveldb::ReadOptions());
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexBlockLink` to protect `leveldb::Iterator* it = m_blockLinkDB->GetDB()->NewIterator(leveldb::ReadOptions());` only

File: src/libPersistence/BlockStorage.cpp [hunk 7]
```diff
@@ -588,7 +684,7 @@ bool BlockStorage::PutDSCommittee(const shared_ptr<DequeOfNode>& dsCommittee,
                                   const uint16_t& consensusLeaderID) {
   LOG_MARKER();
 
-  lock_guard<mutex> g(m_mutexDsCommittee);
+  unique_lock<shared_timed_mutex> g(m_mutexDsCommittee);
   m_dsCommitteeDB->ResetDB();
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Can call `BlockStorage::ResetDB(DBTYPE)` instead

File: src/libPersistence/BlockStorage.cpp [hunk 8]
```diff
@@ -628,15 +724,20 @@ bool BlockStorage::GetDSCommittee(shared_ptr<DequeOfNode>& dsCommittee,
   LOG_MARKER();
 
   unsigned int index = 0;
-  lock_guard<mutex> g(m_mutexDsCommittee);
+  shared_lock<shared_timed_mutex> g(m_mutexDsCommittee);
   string strConsensusLeaderID = m_dsCommitteeDB->Lookup(index++);
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexDsCommittee` to protect `string strConsensusLeaderID = m_dsCommitteeDB->Lookup(index++);` only

File: src/libPersistence/BlockStorage.cpp [hunk 9]
```diff
@@ -662,7 +763,7 @@ bool BlockStorage::PutShardStructure(const DequeOfShard& shards,
                                      const uint32_t myshardId) {
   LOG_MARKER();
 
-  lock_guard<mutex> g(m_mutexShardStructure);
+  unique_lock<shared_timed_mutex> g(m_mutexShardStructure);
   m_shardStructureDB->ResetDB();
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Can call `BlockStorage::ResetDB(DBTYPE)` instead

File: src/libPersistence/BlockStorage.cpp [hunk 10]
```diff
@@ -721,6 +822,8 @@ bool BlockStorage::PutStateDelta(const uint64_t& finalBlockNum,
                                  const bytes& stateDelta) {
   LOG_MARKER();
 
+  unique_lock<shared_timed_mutex> g(m_mutexStateDelta);
+
   if (0 != m_stateDeltaDB->Insert(finalBlockNum, stateDelta)) {
```
Comment (ckyang) [inline src/libPersistence/BlockStorage.cpp]: Maybe let `m_mutexStateDelta` to protect `if (0 != m_stateDeltaDB->Insert(finalBlockNum, stateDelta)) {` only

File: src/libPersistence/BlockStorage.cpp (hunk 1 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 2 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 3 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 4 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 5 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 6 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: cannot do that...the iterator is keep changing in the for loop

File: src/libPersistence/BlockStorage.cpp (hunk 7 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: Will put in next pr

File: src/libPersistence/BlockStorage.cpp (hunk 8 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: Nope... notice that there is a while loop below

File: src/libPersistence/BlockStorage.cpp (hunk 9 above)
Comment (kaikawaliu) [reply src/libPersistence/BlockStorage.cpp]: Not that trivial, will put in the next pr in the future.

**Pull request merged**